File: parser.cpp
// system includes
#include <map>
#include <memory>

// includes
#include "parser.h"
#include "shopndrop_protocol/parser.h"
#include "generic_protocol/parser.h"
#include "basic_objects/parser.h"
#include "basic_parser/parser.h"
#include "basic_parser/malformed_request.h"
#include "validator.h"
#include "request_type_parser.h"

namespace shopndrop_web_protocol
{

namespace parser
{

using ::basic_parser::parser::get_value_or_throw;
using ::basic_parser::parser::get_value_or_throw_t;

// enums

// objects

void get_value_or_throw( ProductItemWithId * res, const std::string & prefix, const generic_request::Request & r )
{
    get_value_or_throw( & res->product_item_id, prefix + ".PRODUCT_ITEM_ID", r );
    ::shopndrop_protocol::parser::get_value_or_throw( & res->product_item, prefix + ".PRODUCT_ITEM", r );
}

void get_value_or_throw( ShoppingItemWithProduct * res, const std::string & prefix, const generic_request::Request & r )
{
    ::shopndrop_protocol::parser::get_value_or_throw( & res->shopping_item, prefix + ".SHOPPING_ITEM", r );
    ::shopndrop_protocol::parser::get_value_or_throw( & res->product_item, prefix + ".PRODUCT_ITEM", r );
}

void get_value_or_throw( ShoppingListWithProduct * res, const std::string & prefix, const generic_request::Request & r )
{
    get_value_or_throw_t( & res->items, prefix + ".ITEMS", r, static_cast<void (*)( ShoppingItemWithProduct * , const std::string & , const generic_request::Request & )>( &get_value_or_throw ) ); // Array
}

void get_value_or_throw( ShoppingListWithTotals * res, const std::string & prefix, const generic_request::Request & r )
{
    get_value_or_throw( & res->shopping_list, prefix + ".SHOPPING_LIST", r );
    get_value_or_throw( & res->price, prefix + ".PRICE", r );
    get_value_or_throw( & res->weight, prefix + ".WEIGHT", r );
}

void get_value_or_throw( RideSummaryWithShopper * res, const std::string & prefix, const generic_request::Request & r )
{
    get_value_or_throw( & res->ride_id, prefix + ".RIDE_ID", r );
    ::shopndrop_protocol::parser::get_value_or_throw( & res->ride, prefix + ".RIDE", r );
    get_value_or_throw( & res->shopper_name, prefix + ".SHOPPER_NAME", r );
}

void get_value_or_throw( RideWithId * res, const std::string & prefix, const generic_request::Request & r )
{
    get_value_or_throw( & res->ride_id, prefix + ".RIDE_ID", r );
    ::shopndrop_protocol::parser::get_value_or_throw( & res->ride, prefix + ".RIDE", r );
}

void get_value_or_throw( ShoppingRequestInfo * res, const std::string & prefix, const generic_request::Request & r )
{
    get_value_or_throw( & res->order_id, prefix + ".ORDER_ID", r );
    get_value_or_throw( & res->sum, prefix + ".SUM", r );
    get_value_or_throw( & res->earning, prefix + ".EARNING", r );
    get_value_or_throw( & res->weight, prefix + ".WEIGHT", r );
    ::shopndrop_protocol::parser::get_value_or_throw( & res->address, prefix + ".ADDRESS", r );
}

void get_value_or_throw( AcceptedOrderUser * res, const std::string & prefix, const generic_request::Request & r )
{
    get_value_or_throw( & res->order_id, prefix + ".ORDER_ID", r );
    ::basic_objects::parser::get_value_or_throw( & res->delivery_time, prefix + ".DELIVERY_TIME", r );
    ::shopndrop_protocol::parser::get_value_or_throw( & res->order, prefix + ".ORDER", r );
    get_value_or_throw( & res->sum, prefix + ".SUM", r );
    get_value_or_throw( & res->shopper_name, prefix + ".SHOPPER_NAME", r );
}

void get_value_or_throw( AcceptedOrderShopper * res, const std::string & prefix, const generic_request::Request & r )
{
    get_value_or_throw( & res->order_id, prefix + ".ORDER_ID", r );
    ::basic_objects::parser::get_value_or_throw( & res->delivery_time, prefix + ".DELIVERY_TIME", r );
    ::shopndrop_protocol::parser::get_value_or_throw( & res->order, prefix + ".ORDER", r );
    get_value_or_throw( & res->sum, prefix + ".SUM", r );
    get_value_or_throw( & res->earning, prefix + ".EARNING", r );
    get_value_or_throw( & res->weight, prefix + ".WEIGHT", r );
}

void get_value_or_throw( DashScreenUser * res, const std::string & prefix, const generic_request::Request & r )
{
    ::basic_objects::parser::get_value_or_throw( & res->current_time, prefix + ".CURRENT_TIME", r );
    get_value_or_throw_t( & res->rides, prefix + ".RIDES", r, static_cast<void (*)( RideSummaryWithShopper * , const std::string & , const generic_request::Request & )>( &get_value_or_throw ) ); // Array
    get_value_or_throw_t( & res->orders, prefix + ".ORDERS", r, static_cast<void (*)( AcceptedOrderUser * , const std::string & , const generic_request::Request & )>( &get_value_or_throw ) ); // Array
}

void get_value_or_throw( DashScreenShopper * res, const std::string & prefix, const generic_request::Request & r )
{
    ::basic_objects::parser::get_value_or_throw( & res->current_time, prefix + ".CURRENT_TIME", r );
    get_value_or_throw_t( & res->rides, prefix + ".RIDES", r, static_cast<void (*)( RideWithId * , const std::string & , const generic_request::Request & )>( &get_value_or_throw ) ); // Array
    get_value_or_throw_t( & res->orders, prefix + ".ORDERS", r, static_cast<void (*)( AcceptedOrderShopper * , const std::string & , const generic_request::Request & )>( &get_value_or_throw ) ); // Array
}

// base messages

// messages

void get_value_or_throw( GetProductItemListRequest * res, const generic_request::Request & r )
{
    // base class
    ::shopndrop_protocol::parser::get_value_or_throw( static_cast<shopndrop_protocol::Request*>( res ), r );

}

void get_value_or_throw( GetProductItemListResponse * res, const generic_request::Request & r )
{
    // base class
    ::generic_protocol::parser::get_value_or_throw( static_cast<generic_protocol::BackwardMessage*>( res ), r );

    get_value_or_throw_t( & res->product_items, "PRODUCT_ITEMS", r, static_cast<void (*)( ProductItemWithId * , const std::string & , const generic_request::Request & )>( &get_value_or_throw ) ); // Array
}

void get_value_or_throw( GetShoppingRequestInfoRequest * res, const generic_request::Request & r )
{
    // base class
    ::shopndrop_protocol::parser::get_value_or_throw( static_cast<shopndrop_protocol::Request*>( res ), r );

    get_value_or_throw( & res->ride_id, "RIDE_ID", r );
}

void get_value_or_throw( GetShoppingRequestInfoResponse * res, const generic_request::Request & r )
{
    // base class
    ::generic_protocol::parser::get_value_or_throw( static_cast<generic_protocol::BackwardMessage*>( res ), r );

    get_value_or_throw_t( & res->requests, "REQUESTS", r, static_cast<void (*)( ShoppingRequestInfo * , const std::string & , const generic_request::Request & )>( &get_value_or_throw ) ); // Array
}

void get_value_or_throw( GetShoppingListWithTotalsRequest * res, const generic_request::Request & r )
{
    // base class
    ::shopndrop_protocol::parser::get_value_or_throw( static_cast<shopndrop_protocol::Request*>( res ), r );

    get_value_or_throw( & res->shopping_list_id, "SHOPPING_LIST_ID", r );
}

void get_value_or_throw( GetShoppingListWithTotalsResponse * res, const generic_request::Request & r )
{
    // base class
    ::generic_protocol::parser::get_value_or_throw( static_cast<generic_protocol::BackwardMessage*>( res ), r );

    get_value_or_throw( & res->shopping_list, "SHOPPING_LIST", r );
}

void get_value_or_throw( GetDashScreenUserRequest * res, const generic_request::Request & r )
{
    // base class
    ::shopndrop_protocol::parser::get_value_or_throw( static_cast<shopndrop_protocol::Request*>( res ), r );

    get_value_or_throw( & res->user_id, "USER_ID", r );
    ::shopndrop_protocol::parser::get_value_or_throw( & res->position, "POSITION", r );
}

void get_value_or_throw( GetDashScreenUserResponse * res, const generic_request::Request & r )
{
    // base class
    ::generic_protocol::parser::get_value_or_throw( static_cast<generic_protocol::BackwardMessage*>( res ), r );

    get_value_or_throw( & res->dash_screen, "DASH_SCREEN", r );
}

void get_value_or_throw( GetDashScreenShopperRequest * res, const generic_request::Request & r )
{
    // base class
    ::shopndrop_protocol::parser::get_value_or_throw( static_cast<shopndrop_protocol::Request*>( res ), r );

    get_value_or_throw( & res->user_id, "USER_ID", r );
    ::shopndrop_protocol::parser::get_value_or_throw( & res->position, "POSITION", r );
}

void get_value_or_throw( GetDashScreenShopperResponse * res, const generic_request::Request & r )
{
    // base class
    ::generic_protocol::parser::get_value_or_throw( static_cast<generic_protocol::BackwardMessage*>( res ), r );

    get_value_or_throw( & res->dash_screen, "DASH_SCREEN", r );
}

// to object

Object * to_GetProductItemListRequest( const generic_request::Request & r )
{
    std::unique_ptr<GetProductItemListRequest> res( new GetProductItemListRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetProductItemListResponse( const generic_request::Request & r )
{
    std::unique_ptr<GetProductItemListResponse> res( new GetProductItemListResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetShoppingRequestInfoRequest( const generic_request::Request & r )
{
    std::unique_ptr<GetShoppingRequestInfoRequest> res( new GetShoppingRequestInfoRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetShoppingRequestInfoResponse( const generic_request::Request & r )
{
    std::unique_ptr<GetShoppingRequestInfoResponse> res( new GetShoppingRequestInfoResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetShoppingListWithTotalsRequest( const generic_request::Request & r )
{
    std::unique_ptr<GetShoppingListWithTotalsRequest> res( new GetShoppingListWithTotalsRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetShoppingListWithTotalsResponse( const generic_request::Request & r )
{
    std::unique_ptr<GetShoppingListWithTotalsResponse> res( new GetShoppingListWithTotalsResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetDashScreenUserRequest( const generic_request::Request & r )
{
    std::unique_ptr<GetDashScreenUserRequest> res( new GetDashScreenUserRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetDashScreenUserResponse( const generic_request::Request & r )
{
    std::unique_ptr<GetDashScreenUserResponse> res( new GetDashScreenUserResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetDashScreenShopperRequest( const generic_request::Request & r )
{
    std::unique_ptr<GetDashScreenShopperRequest> res( new GetDashScreenShopperRequest );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

Object * to_GetDashScreenShopperResponse( const generic_request::Request & r )
{
    std::unique_ptr<GetDashScreenShopperResponse> res( new GetDashScreenShopperResponse );

    get_value_or_throw( res.get(), r );

    validator::validate( * res );

    return res.release();
}

// to forward message

basic_parser::Object* to_forward_message( const generic_request::Request & r )
{
    auto type = detect_request_type( r );

    typedef request_type_e KeyType;

    typedef Object* (*PPMF)( const generic_request::Request & r );

#define HANDLER_MAP_ENTRY(_v)       { KeyType::_v,    & to_##_v }

    static const std::map<KeyType, PPMF> funcs =
    {
        HANDLER_MAP_ENTRY( GetProductItemListRequest ),
        HANDLER_MAP_ENTRY( GetProductItemListResponse ),
        HANDLER_MAP_ENTRY( GetShoppingRequestInfoRequest ),
        HANDLER_MAP_ENTRY( GetShoppingRequestInfoResponse ),
        HANDLER_MAP_ENTRY( GetShoppingListWithTotalsRequest ),
        HANDLER_MAP_ENTRY( GetShoppingListWithTotalsResponse ),
        HANDLER_MAP_ENTRY( GetDashScreenUserRequest ),
        HANDLER_MAP_ENTRY( GetDashScreenUserResponse ),
        HANDLER_MAP_ENTRY( GetDashScreenShopperRequest ),
        HANDLER_MAP_ENTRY( GetDashScreenShopperResponse ),
    };

#undef HANDLER_MAP_ENTRY

    auto it = funcs.find( type );

    if( it != funcs.end() )
        return it->second( r );

    return nullptr;
}

using basic_parser::MalformedRequest;

request_type_e  detect_request_type( const generic_request::Request & r )
{
    std::string cmd;

    if( r.get_value( "CMD", cmd ) == false )
        throw MalformedRequest( "CMD is not defined" );

    return RequestTypeParser::to_request_type( cmd );
}

} // namespace parser

} // namespace shopndrop_web_protocol

File: parser.php
<?php

namespace shopndrop_web_protocol;


// includes
require_once __DIR__.'/../shopndrop_protocol/parser.php';
require_once __DIR__.'/../generic_protocol/parser.php';
require_once __DIR__.'/../basic_objects/parser.php';
require_once __DIR__.'/../basic_parser/parser.php';

// enums

// objects

function parse__ProductItemWithId( & $csv_arr, & $offset )
{
    $res = new ProductItemWithId;

    $res->product_item_id = \basic_parser\parse__int( $csv_arr, $offset );
    $res->product_item = \shopndrop_protocol\parse__ProductItem( $csv_arr, $offset );

    return $res;
}

function parse__ShoppingItemWithProduct( & $csv_arr, & $offset )
{
    $res = new ShoppingItemWithProduct;

    $res->shopping_item = \shopndrop_protocol\parse__ShoppingItem( $csv_arr, $offset );
    $res->product_item = \shopndrop_protocol\parse__ProductItem( $csv_arr, $offset );

    return $res;
}

function parse__ShoppingListWithProduct( & $csv_arr, & $offset )
{
    $res = new ShoppingListWithProduct;

    $res->items = \basic_parser\parse__vector( $csv_arr, $offset, '\shopndrop_web_protocol\parse__ShoppingItemWithProduct' ); // Array

    return $res;
}

function parse__ShoppingListWithTotals( & $csv_arr, & $offset )
{
    $res = new ShoppingListWithTotals;

    $res->shopping_list = parse__ShoppingListWithProduct( $csv_arr, $offset );
    $res->price = \basic_parser\parse__float( $csv_arr, $offset );
    $res->weight = \basic_parser\parse__float( $csv_arr, $offset );

    return $res;
}

function parse__RideSummaryWithShopper( & $csv_arr, & $offset )
{
    $res = new RideSummaryWithShopper;

    $res->ride_id = \basic_parser\parse__int( $csv_arr, $offset );
    $res->ride = \shopndrop_protocol\parse__RideSummary( $csv_arr, $offset );
    $res->shopper_name = \basic_parser\parse__string( $csv_arr, $offset );

    return $res;
}

function parse__RideWithId( & $csv_arr, & $offset )
{
    $res = new RideWithId;

    $res->ride_id = \basic_parser\parse__int( $csv_arr, $offset );
    $res->ride = \shopndrop_protocol\parse__Ride( $csv_arr, $offset );

    return $res;
}

function parse__ShoppingRequestInfo( & $csv_arr, & $offset )
{
    $res = new ShoppingRequestInfo;

    $res->order_id = \basic_parser\parse__int( $csv_arr, $offset );
    $res->sum = \basic_parser\parse__float( $csv_arr, $offset );
    $res->earning = \basic_parser\parse__float( $csv_arr, $offset );
    $res->weight = \basic_parser\parse__float( $csv_arr, $offset );
    $res->address = \shopndrop_protocol\parse__Address( $csv_arr, $offset );

    return $res;
}

function parse__AcceptedOrderUser( & $csv_arr, & $offset )
{
    $res = new AcceptedOrderUser;

    $res->order_id = \basic_parser\parse__int( $csv_arr, $offset );
    $res->delivery_time = \basic_objects\parse__LocalTime( $csv_arr, $offset );
    $res->order = \shopndrop_protocol\parse__Order( $csv_arr, $offset );
    $res->sum = \basic_parser\parse__float( $csv_arr, $offset );
    $res->shopper_name = \basic_parser\parse__string( $csv_arr, $offset );

    return $res;
}

function parse__AcceptedOrderShopper( & $csv_arr, & $offset )
{
    $res = new AcceptedOrderShopper;

    $res->order_id = \basic_parser\parse__int( $csv_arr, $offset );
    $res->delivery_time = \basic_objects\parse__LocalTime( $csv_arr, $offset );
    $res->order = \shopndrop_protocol\parse__Order( $csv_arr, $offset );
    $res->sum = \basic_parser\parse__float( $csv_arr, $offset );
    $res->earning = \basic_parser\parse__float( $csv_arr, $offset );
    $res->weight = \basic_parser\parse__float( $csv_arr, $offset );

    return $res;
}

function parse__DashScreenUser( & $csv_arr, & $offset )
{
    $res = new DashScreenUser;

    $res->current_time = \basic_objects\parse__LocalTime( $csv_arr, $offset );
    $res->rides = \basic_parser\parse__vector( $csv_arr, $offset, '\shopndrop_web_protocol\parse__RideSummaryWithShopper' ); // Array
    $res->orders = \basic_parser\parse__vector( $csv_arr, $offset, '\shopndrop_web_protocol\parse__AcceptedOrderUser' ); // Array

    return $res;
}

function parse__DashScreenShopper( & $csv_arr, & $offset )
{
    $res = new DashScreenShopper;

    $res->current_time = \basic_objects\parse__LocalTime( $csv_arr, $offset );
    $res->rides = \basic_parser\parse__vector( $csv_arr, $offset, '\shopndrop_web_protocol\parse__RideWithId' ); // Array
    $res->orders = \basic_parser\parse__vector( $csv_arr, $offset, '\shopndrop_web_protocol\parse__AcceptedOrderShopper' ); // Array

    return $res;
}

// base messages

// messages

function parse__GetProductItemListRequest( & $csv_arr )
{
    $res = new GetProductItemListRequest;

    $offset = 1;

    // base class
    \shopndrop_protocol\parse__Request( $res, $csv_arr, $offset );


    return $res;
}

function parse__GetProductItemListResponse( & $csv_arr )
{
    $res = new GetProductItemListResponse;

    $offset = 1;

    // base class
    \generic_protocol\parse__BackwardMessage( $res, $csv_arr, $offset );

    $res->product_items = \basic_parser\parse__vector( $csv_arr, $offset, '\shopndrop_web_protocol\parse__ProductItemWithId' ); // Array

    return $res;
}

function parse__GetShoppingRequestInfoRequest( & $csv_arr )
{
    $res = new GetShoppingRequestInfoRequest;

    $offset = 1;

    // base class
    \shopndrop_protocol\parse__Request( $res, $csv_arr, $offset );

    $res->ride_id = \basic_parser\parse__int( $csv_arr, $offset );

    return $res;
}

function parse__GetShoppingRequestInfoResponse( & $csv_arr )
{
    $res = new GetShoppingRequestInfoResponse;

    $offset = 1;

    // base class
    \generic_protocol\parse__BackwardMessage( $res, $csv_arr, $offset );

    $res->requests = \basic_parser\parse__vector( $csv_arr, $offset, '\shopndrop_web_protocol\parse__ShoppingRequestInfo' ); // Array

    return $res;
}

function parse__GetShoppingListWithTotalsRequest( & $csv_arr )
{
    $res = new GetShoppingListWithTotalsRequest;

    $offset = 1;

    // base class
    \shopndrop_protocol\parse__Request( $res, $csv_arr, $offset );

    $res->shopping_list_id = \basic_parser\parse__int( $csv_arr, $offset );

    return $res;
}

function parse__GetShoppingListWithTotalsResponse( & $csv_arr )
{
    $res = new GetShoppingListWithTotalsResponse;

    $offset = 1;

    // base class
    \generic_protocol\parse__BackwardMessage( $res, $csv_arr, $offset );

    $res->shopping_list = parse__ShoppingListWithTotals( $csv_arr, $offset );

    return $res;
}

function parse__GetDashScreenUserRequest( & $csv_arr )
{
    $res = new GetDashScreenUserRequest;

    $offset = 1;

    // base class
    \shopndrop_protocol\parse__Request( $res, $csv_arr, $offset );

    $res->user_id = \basic_parser\parse__int( $csv_arr, $offset );
    $res->position = \shopndrop_protocol\parse__GeoPosition( $csv_arr, $offset );

    return $res;
}

function parse__GetDashScreenUserResponse( & $csv_arr )
{
    $res = new GetDashScreenUserResponse;

    $offset = 1;

    // base class
    \generic_protocol\parse__BackwardMessage( $res, $csv_arr, $offset );

    $res->dash_screen = parse__DashScreenUser( $csv_arr, $offset );

    return $res;
}

function parse__GetDashScreenShopperRequest( & $csv_arr )
{
    $res = new GetDashScreenShopperRequest;

    $offset = 1;

    // base class
    \shopndrop_protocol\parse__Request( $res, $csv_arr, $offset );

    $res->user_id = \basic_parser\parse__int( $csv_arr, $offset );
    $res->position = \shopndrop_protocol\parse__GeoPosition( $csv_arr, $offset );

    return $res;
}

function parse__GetDashScreenShopperResponse( & $csv_arr )
{
    $res = new GetDashScreenShopperResponse;

    $offset = 1;

    // base class
    \generic_protocol\parse__BackwardMessage( $res, $csv_arr, $offset );

    $res->dash_screen = parse__DashScreenShopper( $csv_arr, $offset );

    return $res;
}

// generic

class Parser extends \shopndrop_protocol\Parser
{

protected static function parse_csv_array( $csv_arr )
{
    if( sizeof( $csv_arr ) < 1 )
        return self::create_parse_error();

    $handler_map = array(
        // messages
        'shopndrop_web/GetProductItemListRequest'         => 'parse__GetProductItemListRequest',
        'shopndrop_web/GetProductItemListResponse'         => 'parse__GetProductItemListResponse',
        'shopndrop_web/GetShoppingRequestInfoRequest'         => 'parse__GetShoppingRequestInfoRequest',
        'shopndrop_web/GetShoppingRequestInfoResponse'         => 'parse__GetShoppingRequestInfoResponse',
        'shopndrop_web/GetShoppingListWithTotalsRequest'         => 'parse__GetShoppingListWithTotalsRequest',
        'shopndrop_web/GetShoppingListWithTotalsResponse'         => 'parse__GetShoppingListWithTotalsResponse',
        'shopndrop_web/GetDashScreenUserRequest'         => 'parse__GetDashScreenUserRequest',
        'shopndrop_web/GetDashScreenUserResponse'         => 'parse__GetDashScreenUserResponse',
        'shopndrop_web/GetDashScreenShopperRequest'         => 'parse__GetDashScreenShopperRequest',
        'shopndrop_web/GetDashScreenShopperResponse'         => 'parse__GetDashScreenShopperResponse',
    );

    $type = $csv_arr[0][0];

    if( array_key_exists( $type, $handler_map ) )
    {
        $func = '\\shopndrop_web_protocol\\' . $handler_map[ $type ];
        return $func( $csv_arr[0] );
    }

    return \shopndrop_protocol\Parser::parse_csv_array( $csv_arr );
}

}

# namespace_end shopndrop_web_protocol


?>
